### PR TITLE
Add REST-API client and other utilites based on the requests library

### PR DESCRIPTION
**Pull request by torotil**

### CODE REVIEW

File: moflask/requests/auth.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class AuthAppClient(rest.Client):
+    """Client for the impact-stack auth app."""
+
+    @classmethod
+    def from_config(cls):
+        """Create a new auth-app client instance from the app config."""
+        return cls(flask.current_app.config.get("IMPACT_STACK_AUTH_APP_URL"))
+
+    def get_token(self, organization):
+        """Get a JWT token for a sub-organization."""
```
Comment (katzmo) [inline moflask/requests/auth.py:20]: I do find "sub-organization" a bit confusing: sub of what, the AUTH app instance? :thinking:

File: moflask/requests/rest.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+class Client:
+    """A General purpose REST / JSON API-client."""
+
+    def __init__(self, base_url: str, auth=None, request_timeout=2):
+        """Create a new API client instance."""
+        self._session = sessions.Session()
+        self._session.timeout = request_timeout
+        if auth:
+            self._session.auth = auth
```
Comment (katzmo) [inline moflask/requests/rest.py:19]: What could `auth` be?

File: moflask/requests/auth.py (hunk 1 above)
Comment (torotil) [reply moflask/requests/auth.py:20]: Of impact-stack, which is the top-level organization.

File: moflask/requests/__init__.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+"""Useful wrappers for requests."""
+
+# Also provide exceptions to make them available with a single import.
+from requests import exceptions  # pylint: disable=unused-import
+
+from . import auth, sessions
+
+Session = sessions.Session
+__all__ = ["auth", "exceptions", "Session"]
```
[the remark this answers is not in the document]
Comment (torotil) [reply moflask/requests/__init__.py:9]: This is about what’s available if you just import `moflask.requests`. The idea was that we’d import `moflask.requests.rest` explicitly.

File: moflask/requests/rest.py (hunk 2 above)
Comment (torotil) [reply moflask/requests/rest.py:19]: The type signature of auth would be `Callable[[requests.PreparedRequest], requests.PreparedRequest]`. So the simplest one would be:

```
def auth_middleware(request):
    return request
```

File: moflask/requests/auth.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class AuthAppMiddleware:
+    """Middleware for authenticating using JWT tokens.
+
+    The middleware transparently requests an API-token from the auth-app on-demand.
+    """
+
+    def __init__(self, client, organization):
+        """Create new auth-app requests auth middleware."""
+        self.client = client
```
[the remark this answers is not in the document]
Comment (torotil) [reply moflask/requests/auth.py:35]: Good idea.

File: moflask/requests/auth.py (hunk 1 above)
Comment (katzmo) [reply moflask/requests/auth.py:20]: What do I do if I need a token for the organization the API key is for?

File: moflask/requests/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (katzmo) [reply moflask/requests/__init__.py:9]: OK, also fine :)

File: moflask/requests/auth.py (hunk 1 above)
Comment (torotil) [reply moflask/requests/auth.py:20]: > What do I do if I need a token for the organization the API key is for?

Then you have find a new use-case.

Comment (katzmo) [reply moflask/requests/auth.py:20]: This sounds like the top-level organization is something different to what we normally call an "organization"? In that case I think the comment should just say "organization" :wink:

Comment (torotil) [reply moflask/requests/auth.py:20]: I think it’s perfectly fine within the domain language we have within impact-stack / auth-app. At the moment we have  impact-stack as the top-level *organization* and all clients as *sub-organizations* and chapters as *sub-sub-organizations* …

Comment (katzmo) [reply moflask/requests/auth.py:20]: Ok, I've never heard of it before :shrug:

**Pull request merged**